applications/SwimmingDEMApplication/custom_strategies/strategies/bdf2_turbulent_schemeDEMCoupled.h

```diff
@@ -0,0 +1,351 @@
+    typedef typename TDenseSpace::VectorType LocalSystemVectorType;
+
+    typedef Dof<TDataType> TDofType;
+    typedef typename BaseType::DofsArrayType DofsArrayType;
+
+    typedef CoordinateTransformationUtils<LocalSystemMatrixType, LocalSystemVectorType, double> RotationToolType;
+    typedef typename RotationToolType::UniquePointer RotationToolPointerType;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default constructor.
+    BDF2TurbulentSchemeDEMCoupled()
+    : BDF2TurbulentScheme<TSparseSpace, TDenseSpace>()
+    , mrPeriodicIdVar(Kratos::Variable<int>::StaticObject())
+    {}
+
+    /// Constructor to use the formulation combined with a turbulence model.
+    /**
+     * The turbulence model is assumed to be implemented as a Kratos::Process.
+     * The model's Execute() method wil be called at the start of each
+     * non-linear iteration.
+     * @param pTurbulenceModel pointer to the turbulence model
+     */
+    BDF2TurbulentSchemeDEMCoupled(Process::Pointer pTurbulenceModel)
+        : BDF2TurbulentScheme<TSparseSpace, TDenseSpace>()
+        , mpTurbulenceModel(pTurbulenceModel)
+        , mrPeriodicIdVar(Kratos::Variable<int>::StaticObject())
+    {}
+
+    /// Constructor for periodic boundary conditions.
+    /**
+     * @param rPeriodicVar the variable used to store periodic pair indices.
+     */
+    BDF2TurbulentSchemeDEMCoupled(const Kratos::Variable<int>& rPeriodicVar)
+        : BDF2TurbulentScheme<TSparseSpace, TDenseSpace>()
+        , mrPeriodicIdVar(rPeriodicVar)
+    {}
+
+
+    /// Destructor.
+    ~BDF2TurbulentSchemeDEMCoupled() override
+    {}
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /// Set the time iteration coefficients
+    void InitializeSolutionStep(
+        ModelPart& rModelPart,
+        TSystemMatrixType& A,
+        TSystemVectorType& Dx,
+        TSystemVectorType& b) override
+    {
+        ProcessInfo CurrentProcessInfo = rModelPart.GetProcessInfo();
+
+        BDF2TurbulentScheme<TSparseSpace, TDenseSpace>::InitializeSolutionStep(rModelPart, A, Dx, b);
+        this->CalculateFluidFraction(rModelPart, CurrentProcessInfo);
+    }
+
+    void CalculateFluidFraction(
+        ModelPart& r_model_part,
+        ProcessInfo& r_current_process_info)
+    {
+        double delta_time = r_current_process_info[DELTA_TIME];
+        double delta_time_inv = 1.0 / delta_time;
+
+        //for(int k = 0; k<static_cast<int>(rModelPart.Nodes().size()); k++)
+        block_for_each(r_model_part.Nodes(), [&](Node<3>& rNode)
+        {
+            const double fluid_fraction = rNode.FastGetSolutionStepValue(FLUID_FRACTION);
+            const double fluid_fraction_old = rNode.FastGetSolutionStepValue(FLUID_FRACTION_OLD);
+            const double fluid_fraction_rate = delta_time_inv * (fluid_fraction - fluid_fraction_old);
+            rNode.FastGetSolutionStepValue(FLUID_FRACTION_RATE) = fluid_fraction_rate;
```

Comment:
I think this should be made coherent with the bdf2 scheme... It should be easy by copying what is done in InitializeSolutionStep from the original bdf2 scheme

Reply:
This is still not bdf2, right?

Reply:
You have to press F5 to see the changes and go to the files changed in the commit before the last one to see the modifications in the bdf2